### ISSUE TITLE
Unable to connect to ADP UAT environment

### ISSUE THREAD

**Issue by sudiptodesmukh-okta** [adplabs/adp-connection-java#4]
Hi, 
1)      I am successfully able to connect to iat-api.adp.com using the packaged credentials for the Java Developer Library pointing to Client Credentials Flow of OIDC
2)      All I need is to be able to connect to ADP UAT or later ADP Prod using the same way
3)      I’m hitting a roadblock while trying to follow the ONE-TIME_SETUP instructions to configure certs for ADP UAT using attached key and PEM (was provided by Workforce Now Team)
 
K2338:certs sudiptodesmukh$ openssl pkcs12 -export -out keystore_UAT.pkcs12 -in Prometheus_UAT.pem -inkey Prometheus_auth.key
unable to load private key
43591:error:0906D06C:PEM routines:PEM_read_bio:no start line:/BuildRoot/Library/Caches/com.apple.xbs/Sources/OpenSSL098/OpenSSL098-59.60.1/src/crypto/pem/pem_lib.c:648:Expecting: ANY PRIVATE KEY

The workforceNow team also created a JKS for us and on using that, we hit the following issue.
Below are the steps carried out –
1)       Copied the JKS file provided to the certs folder in the SDK
2)       Change clientConfig.properties to reflect Prometheus UAT environment (attached)
3)       Ran the code provided on Prometheus site which retrieves a AccessToken but it’s unable to connect to the environment.
 
Error received :
getADPAuthorization Exception com.adp.marketplace.connection.exception.ConnectionException: com.adp.marketplace.connection.exception.ConnectionException: com.adp.marketplace.connection.exception.ConnectionException: java.security.UnrecoverableKeyException: Cannot recover key
Exception in thread "main" java.lang.NullPointerException
                at com.adp.marketplace.demo.client.auth.clientcred.ADPDemoClientCredentials.main(ADPDemoClientCredentials.java:78)
I'll be happy to share the cert files etc if need be.

**Comment by Mahesh1240** [comment]
Hi,
Are you trying to access the UAT environment using the same certificate which is used for IAT environment? If yes, can you please generate a new certificate for UAT as well and use for the UAT calls.

Also please test your certificate by running the below commands before making the API calls.

Test your certificate. Run the following curl to determine if are properly attaching your signed certificate to your request.
Request:
Option 1 -
curl -v --cert yourCert.pfx:yourCertPassword https://api.adp.com/mirror
Option 2 -
curl -v --cert yourCert.pem --key yourKey.key https://api.adp.com/mirror 


Thanks,
Mahesh.

**Comment by ADPMarketplace** [comment]
Hi,
Are you trying to access the UAT environment using the same certificate which is used for IAT environment? If yes, can you please generate a new certificate for UAT as well and use for the UAT calls.

Also please test your certificate by running the below commands before making the API calls.

Test your certificate. Run the following curl to determine if are properly attaching your signed certificate to your request.
Request:
Option 1 -
curl -v --cert yourCert.pfx:yourCertPassword https://api.adp.com/mirror
Option 2 -
curl -v --cert yourCert.pem --key yourKey.key https://api.adp.com/mirror 

Replied to the user on Github as well.


Thanks,
Mahesh.

**Comment by sudiptodesmukh-okta** [comment]
Hi Mahesh,
I ran the command mentioned by you and got the following output, can you please advise ?

K2338:Downloads sudiptodesmukh$ curl -v --cert Prometheus_UAT.pem --key prometheus_auth.key https://api.adp.com/mirror
*   Trying 170.146.98.12...
* Connected to api.adp.com (170.146.98.12) port 443 (#0)
* WARNING: SSL: CURLOPT_SSLKEY is ignored by Secure Transport. The private key must be in the Keychain.
* WARNING: SSL: Certificate type not set, assuming PKCS#12 format.
* SSL: Can't load the certificate "Prometheus_UAT.pem" and its private key: OSStatus -25299
* Closing connection 0
curl: (58) SSL: Can't load the certificate "Prometheus_UAT.pem" and its private key: OSStatus -25299
K2338:Downloads sudiptodesmukh$

If you like I can send the PEM and the Key File in a Secure Email to you. I didn’t try the PFX file as we need to generate a JKS file from this which although was provided to us doesn’t work.

Thanks
Sudipto Desmukh, [cid:image001.png@01D36AD6.FBEF83C0] <https://www.okta.com/>

**Comment by ADPMarketplace** [comment]
Hi,
Can you please share the .PEM file and password through secure email. Would like to test at my end.

Thanks,
Mahesh.

**Comment by ADPMarketplace** [comment]
Request #55625 "URGENT : Unable to connect to AD..." was closed and merged into this request. Last comment in request #55625:

Hi,
Can you please share the .PEM file and password through secure email. Would like to test at my end.

Thanks,
Mahesh.

**Comment by Mahesh1240** [comment]
Hi,
Also please let us know where are you testing the CURL command?

Thanks,
Mahesh.

**Comment by Mahesh1240** [comment]
Hi,
Can you please reopen the case, if the issue still persists.

I am not sure why it got closed automatically?

Thanks,
Mahesh

**Comment by sudiptodesmukh-okta** [comment]
Testing the cURL command using a Terminal on my Mac, certificate was issue for this system. Btw whats your emailId for me to send the .pem file securely ?

Thanks
Sudipto Desmukh, [cid:image001.png@01D36CED.E8FA40C0] <https://www.okta.com/>

**Comment by Mahesh1240** [comment]
Please email me at "Maheshwara.Pothrakanda@ADP.com".

**Comment by sudiptodesmukh-okta** [comment]
@Mahesh1240 I sent you an email with details on latest issues... pasting it here too, plese reopen this case.

Hello, 
While I was able to generate keystore.jks using the new certs provided, I’m again having issues connecting to UAT environment via a simple REST call.
I have followed all steps as in attached doc to make sure that certs are imported to cacerts. Interestingly, I’m able to connect to IAT environment fine.

Can someone have a screen share to help troubleshoot this ?

Thanks
Sudipto
